modules/twinklespeedy.js

```diff
@@ -998,84 +1027,55 @@ Twinkle.speedy.normalizeHash = {
-	'redundantimage': 'Redundant copy of non-Commons file in the same file format',
-	'noimage': 'Corrupt or empty file',
-	'fpcfail': 'Unneeded file description page for a file on Commons',
-	'noncom': 'Invalid licence, eg. "for non-commercial use only" or "for Wikipedia use only"',
-	'unksource': 'Lack of licensing information',
-	'unfree': 'Unused non-free media',
-	'norat': 'Non-free file with no [[WP:RAT|non-free use rationale]]',
-	'badfairuse': 'Violates [[WP:F|non-free use policy]]',
-	'nowcommons': 'Media file available on Commons',
-	'imgcopyvio': 'Unambiguous [[WP:COPYVIO|copyright violation]]',
-	'badfiletype': 'Useless media file (not an image, audio or video)',
-	'nopermission': 'No evidence of permission',
-// Categories
-	'catempty': 'Empty category',
-// User pages
-	'userreq': 'User request to delete page in own userspace',
-	'nouser': 'Userpage or subpage of a nonexistent user',
-	'gallery': '[[WP:NFC|Non-free]] [[Help:Gallery|gallery]]',
-	'notwebhost': '[[WP:NOTWEBHOST|Misuse of Wikipedia as a web host]]',
-// Templates
-	'policy': 'Template that unambiguously misrepresents established policy',
-	'duplicatetemplate': 'Unused, redundant template',
-// Portals
-	'p1': '[[WP:P|Portal]] page that would be subject to speedy deletion as an article',
-	'emptyportal': '[[WP:P|Portal]] without a substantial topic base',
-// Redirects
-	'rediruser': 'Cross-[[WP:NS|namespace]] [[WP:R|redirect]] from mainspace',
-	'redirtypo': 'Recently created, implausible [[WP:R|redirect]]'
-};
-
 Twinkle.speedy.callbacks = {
+	getTemplateCodeAndParams: function(params) {
+		var code, parameters, i;
+		if (params.normalizeds.length > 1) {
+			code = "{{db-multiple";
+			params.utparams = {};
+			$.each(params.normalizeds, function(index, norm) {
+				code += "|" + norm.toUpperCase();
+				parameters = params.templateParams[index] || [];
+				for (var i in parameters) {
+					if (typeof parameters[i] === 'string' && !parseInt(i, 10)) {  // skip numeric parameters - {{db-multiple}} doesn't understand them
+						code += "|" + i + "=" + parameters[i];
+					}
+				}
+				$.extend(params.utparams, Twinkle.speedy.getUserTalkParameters(norm, parameters));
+			});
+			code += "}}";
+		} else {
+			parameters = params.templateParams[0] || [];
+			code = "{{db-" + params.values[0];
+			for (i in parameters) {
+				if (typeof parameters[i] === 'string') {
+					code += "|" + i + "=" + parameters[i];
+				}
+			}
+			if (params.usertalk) {
+				code += "|help=off";
+			}
+			code += "}}";
+			params.utparams = Twinkle.speedy.getUserTalkParameters(params.normalizeds[0], parameters);
+		}
+
+		return [code, params];
+	},
+
+	parseWikitext: function(wikitext, callback) {
+		var query = {
+			action: "parse",
+			prop: "text",
+			pst: "true",
+			text: wikitext,
+			title: mw.config.get("wgPageName")
+		};
+		var api = new Morebits.wiki.api( 'Fetching deletion summary from template', query, callback);
+		api.post({
+			async: false
```

Comment:
Yeah, I wouldn't normally do async either but I saw it was being done here on [L1647](https://github.com/MusikAnimal/twinkle/blob/speedy-admins/modules/twinklespeedy.js#L1647). It's just so tough to work without deferred/promises! I'll do a series of callbacks instead